pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
+        size_t col_block_size = row_block_size;
+
+        for (size_t row = 0; row < num_rows; row++) {
+            for (size_t col_idx = static_cast<size_t>(csrOffsets_ptr[row]);
+                 col_idx < static_cast<size_t>(csrOffsets_ptr[row + 1]);
+                 col_idx++) {
+
+                size_t current_global_row = row;
+                size_t current_global_col = columns_ptr[col_idx];
+                std::complex<Precision> current_val = values_ptr[col_idx];
+
+                size_t block_row_id = current_global_row / row_block_size;
+                size_t block_col_id = current_global_col / col_block_size;
+
+                size_t local_row_id = current_global_row % row_block_size;
+                size_t local_col_id = current_global_col % col_block_size;
+
+                if (splitSparseMatrix[block_row_id][block_col_id]
+                        .csrOffsets.size() == 0) {
+                    splitSparseMatrix[block_row_id][block_col_id].csrOffsets =
+                        std::vector<index_type>(row_block_size + 1, 0);
+                }
+
+                splitSparseMatrix[block_row_id][block_col_id]
+                    .csrOffsets[local_row_id + 1]++;
+                splitSparseMatrix[block_row_id][block_col_id].columns.push_back(
+                    local_col_id);
+                splitSparseMatrix[block_row_id][block_col_id].values.push_back(
+                    current_val);
+            }
+        }
+
+        for (size_t block_row_id = 0; block_row_id < num_row_blocks;
+             block_row_id++) {
+            for (size_t block_col_id = 0; block_col_id < num_col_blocks;
+                 block_col_id++) {
+                for (size_t i0 = 1;
+                     i0 < splitSparseMatrix[block_row_id][block_col_id]
+                              .csrOffsets.size();
+                     i0++) {
+                    splitSparseMatrix[block_row_id][block_col_id]
+                        .csrOffsets[i0] +=
+                        splitSparseMatrix[block_row_id][block_col_id]
+                            .csrOffsets[i0 - 1];
+                }
+            }
+        }
+
+        return splitSparseMatrix;
+    }
+
+    /**
+     * @brief expval(H) calculation with cuSparseSpMV.
+     *
+     * @tparam index_type Integer type used as indices of the sparse matrix.
+     * @param csr_Offsets_ptr Pointer to the array of row offsets of the sparse
+     * matrix. Array of size csrOffsets_size.
+     * @param csrOffsets_size Number of Row offsets of the sparse matrix.
+     * @param columns_ptr Pointer to the array of column indices of the sparse
+     * matrix. Array of size numNNZ
+     * @param values_ptr Pointer to the array of the non-zero elements
+     * @param numNNZ Number of non-zero elements.
+     * @return auto Expectation value.
+     */
+    template <class index_type>
+    auto getExpectationValueOnSparseSpMV(
+        const index_type *csrOffsets_ptr, const index_type csrOffsets_size,
```

Comment:
We can probably replace the pointers here again. If a pointer is necessary, can we consider using a `std::span` instead?

Reply:
See above.